<a href="https://colab.research.google.com/github/flora0110/sentence_prediction/blob/main/bge_m3_test.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
import json

with open('/content/drive/MyDrive/senior/LAB/LawCaseRetreiver/sentence_predict/dataset_new/train_data.json', 'r', encoding='utf-8') as f:
    train_data = json.load(f)

with open('/content/drive/MyDrive/senior/LAB/LawCaseRetreiver/sentence_predict/dataset_new/test_data.json', 'r', encoding='utf-8') as f:
    test_data = json.load(f)

In [4]:
print("===Train Dataset===")
print("Key: ", train_data[0].keys())
print("Size: ", len(train_data))

print("===Test Dataset===")
print("Key: ", test_data[0].keys())
print("Size: ", len(test_data))

===Train Dataset===
Key:  dict_keys(['id', 'reason', 'sentance', 'fact'])
Size:  500
===Test Dataset===
Key:  dict_keys(['id', 'reason', 'sentance', 'fact'])
Size:  100


In [5]:
train_fact_len = []
test_fact_len = []

for data in train_data:
    train_fact_len.append(len(data['fact']))

for data in test_data:
    test_fact_len.append(len(data['fact']))


In [6]:
print("===Train Dataset===")
print("Length: ", len(train_fact_len))
print("Max Length: ", max(train_fact_len))
print("Min Length: ", min(train_fact_len))
print("Avg. Length: ", sum(train_fact_len) / len(train_fact_len))

print("===Test Dataset===")
print("Length: ", len(test_fact_len))
print("Max Length: ", max(test_fact_len))
print("Min Length: ", min(test_fact_len))
print("Avg. Length: ", sum(test_fact_len) / len(test_fact_len))

===Train Dataset===
Length:  500
Max Length:  19745
Min Length:  328
Avg. Length:  7933.656
===Test Dataset===
Length:  100
Max Length:  19277
Min Length:  280
Avg. Length:  8231.62


In [7]:
def extract_key_contents(data):
    facts = []
    sentences = []
    for item in data:
        facts.append(item.get('fact'))
        sentences.append(item.get('sentance'))
    return facts, sentences

In [8]:
train_facts, train_sentences = extract_key_contents(train_data)
test_facts, test_sentences = extract_key_contents(test_data)

In [9]:
print("Train facts:", train_facts[0])
print("Train Sentences:", train_sentences[:5])
print("Test facts:", test_facts[0])
print("Test Sentences:", test_sentences[:5])

Train facts: 一、陳泰龍明知其與泰國籍女子陳玉婉（泰國名字為JARUWANWONGPRATOOM，陳玉婉涉嫌本案偽造文書部分，另經臺灣臺南地方法院檢察署檢察官通緝中）間並無結婚之真意，然因詹巧薇、張晉源等人欲找臺灣男子前往泰國與泰國女子辦理假結婚，林清水負責介紹臺灣男子予詹巧薇、張晉源等人以賺取新臺幣（下同）1萬元之仲介費，適陳泰龍依林清水於報紙上刊登之徵人廣告應徵時，由林清水告知泰國女子欲以假結婚之方式申請來臺居留工作，其若擔任假結婚之人頭老公，除不用負擔前往泰國之食宿、機票費用外，尚可獲取5、6萬元之報酬，迨陳泰龍應允同意後，林清水將其轉介紹予詹巧薇、張晉源等人，陳泰龍遂與林清水、詹巧薇、張晉源等人共同基於行使使公務員登載不實公文書之概括犯意聯絡，先由陳泰龍依指示於民國93年4月3日前往泰國，並於同年月7日與陳玉婉在泰國曼谷市法拉坎農區註冊辦事處登記結婚，取得上開註冊辦事處核發之結婚登記書、結婚證書等文件後，陳泰龍即行返臺，並於同年5月19日持經我國駐泰國臺北經濟文化辦事處認證之上開結婚登記書、結婚證書等資料，至臺南縣永康市（改制前稱謂，以下沿用舊制稱謂）戶政事務所填寫結婚登記申請書，申請辦理其與陳玉婉之結婚登記，因此使上開戶政事務所僅有形式審查權之承辦公務員陷於錯誤，於形式上審核後，將陳泰龍與陳玉婉於93年4月7日結婚之不實事項，登載於其職務上所掌屬公文書之戶籍登記資料內，並據以核發戶籍謄本，而足以生損害於戶政機關對於戶政管理及戶籍登記資料維護之正確性。迨陳玉婉來臺後，為取得居留身分，再由陳泰龍、陳玉婉承前概括之犯意聯絡，於同年6月14日及94年5月5日，均持前開登載不實之戶籍謄本等資料，先後向臺南縣政府警察局外事課申請核發陳玉婉之外僑居留證及申請居留證延期而連續行使之（業經檢察官當庭補充），經承辦人員實質審查結果，均未發覺陳泰龍與陳玉婉假結婚之實情，而發給陳玉婉居留證或核准延期，各足以生損害於主管機關對外僑居留管理之正確性。二、案經臺南市警察局（現改制為臺南市政府警察局，以下沿用舊制稱謂）第一分局報告臺灣臺南地方法院檢察署檢察官偵查起訴。理由甲、程序方面：一、本件以下所引用具傳聞證據性質之供述證據，因本件被告陳泰龍、檢察官等人於本院準備程序進行時，均未爭執前開陳述之證據能力，且迄至本案言詞辯論終結前亦均未聲明異議，應視為已同意援引作為證據，又本院

# 斷詞

In [10]:
!pip install udicOpenData

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.8/17.8 MB 38.6 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for udicOpenData: filename=udicOpenData-2.4-py3-none-any.whl size=18094423 sha256=68028ea27335beadbf2a842d44509d0408aa77bcec05aa1d2979f27ac28d7f38
  Stored in directory: /root/.cache/pip/wheels/86/4f/46/2a28bc28485ba1f9d67dead62ac6aba14089a6cf1d5fbad365
Successfully built udicOpenData


In [11]:
!pip install opencc-python-reimplemented

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 481.8/481.8 kB 3.3 MB/s eta 0:00:00


In [12]:
from opencc import OpenCC

In [13]:
cc = OpenCC('t2s')

In [14]:
from udicOpenData.stopwords import *

# 分詞加空白
def split_passage(passage):
    word_and_tag = list(rmsw(passage, flag=True))
    _temp = list(filter(lambda x: x[1]!='nr' and x[1]!='ns', word_and_tag))
    # print(_temp)
    text = " ".join(list(map(lambda x:x[0], _temp)))
    return text

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger.zip.
[nltk_data] Downloading package maxent_ne_chunker to
[nltk_data]     /root/nltk_data...
[nltk_data]   Unzipping chunkers/maxent_ne_chunker.zip.
[nltk_data] Downloading package words to /root/nltk_data...
[nltk_data]   Unzipping corpora/words.zip.
[nltk_data] Downloading package wordnet to /root/nltk_data...


# bge-m3

In [18]:
!pip install -U FlagEmbedding

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 120.4/120.4 kB 1.6 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 510.5/510.5 kB 6.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 290.1/290.1 kB 21.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 156.5/156.5 kB 18.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 23.7/23.7 MB 37.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 823.6/823.6 kB 51.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.1/14.1 MB 40.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 731.7/731.7 MB 519.8 kB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 410.6/410.6 MB 824.7 kB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 121.6/121.6 MB 1.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.5/56.5 MB 8.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 124.2/12

In [19]:
from FlagEmbedding import BGEM3FlagModel

In [20]:
model = BGEM3FlagModel('BAAI/bge-m3',
                       use_fp16=True) # Setting use_fp16 to True speeds up computation with a slight performance degradation

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:88: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


Fetching 23 files:   0%|          | 0/23 [00:00<?, ?it/s]

.gitattributes:   0%|          | 0.00/1.57k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/687 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/191 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/15.5k [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/123 [00:00<?, ?B/s]

imgs/bm25.jpg:   0%|          | 0.00/69.0k [00:00<?, ?B/s]

imgs/.DS_Store:   0%|          | 0.00/6.15k [00:00<?, ?B/s]

imgs/mkqa.jpg:   0%|          | 0.00/608k [00:00<?, ?B/s]

imgs/others.webp:   0%|          | 0.00/21.0k [00:00<?, ?B/s]

long.jpg:   0%|          | 0.00/127k [00:00<?, ?B/s]

imgs/nqa.jpg:   0%|          | 0.00/158k [00:00<?, ?B/s]

modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

colbert_linear.pt:   0%|          | 0.00/2.10M [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/54.0 [00:00<?, ?B/s]

imgs/long.jpg:   0%|          | 0.00/485k [00:00<?, ?B/s]

imgs/miracl.jpg:   0%|          | 0.00/448k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/964 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/1.31k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/2.27G [00:00<?, ?B/s]

sentencepiece.bpe.model:   0%|          | 0.00/5.07M [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/2.27G [00:00<?, ?B/s]

sparse_linear.pt:   0%|          | 0.00/3.52k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/17.1M [00:00<?, ?B/s]

# TEST
- 對角線都應該越大越好

In [16]:
sentences_1 = ["What is BGE M3?", "Defination of BM25"]
sentences_2 = ["BGE M3 is an embedding model supporting dense retrieval, lexical matching and multi-vector interaction.",
               "BM25 is a bag-of-words retrieval function that ranks a set of documents based on the query terms appearing in each document"]

繁體詞測試

In [17]:
tword_1 = ["老師", "商品"]
tword_2 = ["教師", "貨物"]

簡體詞測試

In [21]:
sword_1 = ["老师", "商品"]
sword_2 = ["教师", "货物"]

sample_2[0] 為 sample_1[0] 的 summary

sample_2[1] 為 sample_1[1] 的 summary

In [22]:
sample_1 = ["陳泰龍明知其與泰國籍女子陳玉婉（泰國名字為JARUWANWONGPRATOOM，陳玉婉涉嫌本案偽造文書部分，另經臺灣臺南地方法院檢察署檢察官通緝中）間並無結婚之真意，然因詹巧薇、張晉源等人欲找臺灣男子前往泰國與泰國女子辦理假結婚，林清水負責介紹臺灣男子予詹巧薇、張晉源等人以賺取新臺幣（下同）1萬元之仲介費，適陳泰龍依林清水於報紙上刊登之徵人廣告應徵時，由林清水告知泰國女子欲以假結婚之方式申請來臺居留工作，其若擔任假結婚之人頭老公，除不用負擔前往泰國之食宿、機票費用外，尚可獲取5、6萬元之報酬，迨陳泰龍應允同意後，林清水將其轉介紹予詹巧薇、張晉源等人，陳泰龍遂與林清水、詹巧薇、張晉源等人共同基於行使使公務員登載不實公文書之概括犯意聯絡，先由陳泰龍依指示於民國93年4月3日前往泰國，並於同年月7日與陳玉婉在泰國曼谷市法拉坎農區註冊辦事處登記結婚，取得上開註冊辦事處核發之結婚登記書、結婚證書等文件後，陳泰龍即行返臺，並於同年5月19日持經我國駐泰國臺北經濟文化辦事處認證之上開結婚登記書、結婚證書等資料，至臺南縣永康市（改制前稱謂，以下沿用舊制稱謂）戶政事務所填寫結婚登記申請書，申請辦理其與陳玉婉之結婚登記，因此使上開戶政事務所僅有形式審查權之承辦公務員陷於錯誤，於形式上審核後，將陳泰龍與陳玉婉於93年4月7日結婚之不實事項，登載於其職務上所掌屬公文書之戶籍登記資料內，並據以核發戶籍謄本，而足以生損害於戶政機關對於戶政管理及戶籍登記資料維護之正確性",
               "張慶堦係址設宜蘭縣羅東鎮○○路289之7號12樓泰雅工程有限公司（以下簡稱泰雅公司）實際負責人，張文騫係址設臺中市○區○○街229號7樓九慶工程顧問有限公司（以下簡稱九慶公司）工程師，九慶公司係承攬宜蘭縣政府文化局發包之「2010宜蘭國際童玩藝術節超級High-net體能極限網設施委託監造服務案」，張文騫為現場監工之人員，吳建忠係宜蘭縣政府文化局約聘人員，為「2010宜蘭國際童玩藝術節超級High-net體能極限網設施委託服務」招標案之承辦人員，均為從事業務之人，泰雅公司於民國99年4月間以新臺幣（下同）1,935萬元得標承作宜蘭縣政府文化局「2010宜蘭國際童玩藝術節超級High-net體能極限網設施委託服務」之標案，依規定於投標時根據招標文件「評選須知」及「設施效果說明及規範」之要求製作「服務建議書」作為投標文件，俾利評選委員審查、評分及作為履約之依據，依據泰雅公司投標時所提出之99年4月份「2010宜蘭國際童玩藝術節超級High-net體能極限網設施結構分析計算書」，有關前開工程之主、副吊橋鋼索尺寸原設計為50mm及40mm，惟於製作「2010宜蘭國際童玩藝術節超級High-net體能極限網超級High-net施作詳圖」時，主吊橋鋼索尺寸註記為60mm，副吊橋鋼索尺寸仍為40mm，泰雅公司於得標後分包予「礁溪鋼鐵機械有限公司」（以下簡稱礁溪公司）施作時，因40mm尺寸之鋼索缺貨，僅剩36mm尺寸之鋼索，嗣張慶堦將主副吊橋鋼索尺寸修正為50mm及36mm，而上開規格經結構技師郭朝和確認安全無虞後，由張慶堦於99年6月14日將經結構技師郭朝和簽證之99年6月份「2010宜蘭國際童玩藝術節超級High-net體能極限網設施結構分析計算書」等文件送請宜蘭縣政府文化局審議，於審議通過後即依審定之相關工程圖說施工，張慶堦實際施工時副吊橋鋼索尺寸以38mm施工（就鋼索尺寸由60mm及40mm更改為50mm及38、36mm並無價差問題），惟宜蘭縣政府文化局於辦理驗收時，發現主副吊橋鋼索尺寸與前開未送審前「2010宜蘭國際童玩藝術節超級High-net體能極限網超級High-net施作詳圖」中所附「橋面鋼索平面圖」所載鋼索尺寸不符，張慶堦遂向宜蘭縣政府文化局驗收人員表示曾於工務會議中決定鋼索尺寸變更一事，宜蘭縣政府文化局於99年9月間泰雅公司欲請領尾款時，要求泰雅公司提出前開工務會議紀錄，張慶堦、張文騫、吳建忠等3人明知本案並未開會討論且未於會議中經結構技師主動建議主、副吊橋鋼索尺寸從60mm、40mm修改為50mm、36mm等情，為利泰雅公司取得工程尾款，張慶堦、張文騫、吳建忠竟基於行使業務上登載不實文書之犯意聯絡，由張慶堦99年9月間某日於其業務上所作成內容為99年5月10日在冬山河工務所召開之會議紀錄，其中決議事項1.虛偽記載「基於安全性考量及為減輕上層爬網重量，經結構技師建議原設計圖為60mm鋼攬（7股19絲），改為50mm（6股37絲）施作，40mm鋼攬改由36mm鋼攬施作，拉力不變及重量減輕」之不實內容，而後於製作完後隔日交由張文騫及吳建忠簽名確認，再由張慶堦向宜蘭縣政府文化局申請工程尾款而行使之，足以生損害於結構技師郭朝和及宜蘭縣政府文化局對於文書管理之正確性"]
sample_2 = ["在一個國家，一名中介人員介紹一名本國男子與外國女子假結婚，以幫助外國女子獲得該國的居留權。中介人員告訴男子，只要他同意假結婚，就可以獲得一筆可觀的報酬，並免費提供旅行和住宿等費用。男子同意了這個提議，並與外國女子在該國的一個地方辦理了假結婚手續。然後，他們向當地政府提出結婚登記申請，該政府在審核文件時被誤導，最終登記了這對假夫妻的結婚信息，導致對當地政府戶政管理的損害。",
               "一家建築公司在承包政府部門的橋樑工程後，發現原定使用的建材價格上漲，導致成本增加。為了避免損失，該公司的管理層決定在最後一次驗收時，提交一份虛偽的施工報告，聲稱他們使用了更高價格的建材以提高橋樑的品質。然而實際上，他們使用了較便宜的建材，並在報告中捏造了資料。這麼做是為了讓政府部門相信工程的品質更高，從而順利取得工程尾款。這樣的行為不僅對工程品質造成損害，也損害了政府部門的信任和資源的分配。"]

sample_3[0] 和 sample_4[0] 為相似案件（假結婚）

sample_3[1] 和 sample_4[1] 為相似案件（冒用他人名義 遺產）

sample_3[2] 和 sample_4[2] 為相似案件（繼承 扭曲事實）

In [64]:
sample_3 = ["己○○與經營「金晉安人力仲介有限公司」之官威成、官怡秀（所涉違反就業服務法等案件，分別經臺灣南投地方法院以95年度易字第53號判處有期徒刑10月，及臺灣高等法院臺中分院以99年度上易字第94號判處有期徒刑6月確定）及真實姓名年籍不詳綽號「李姐」之成年女子，均明知如附表所示之外籍女子傅麗娜等人僅為來臺工作而無結婚之真意，竟由己○○負責在臺灣居中介紹如附表所示亦無結婚真意之戊○○等人擔任人頭老公，渠等即共同基於明知為不實事項而使公務員登載於職務上所掌公文書之概括犯意聯絡，由己○○、官威成、官怡秀及「李姐」等人為擔任人頭老公之戊○○、丙○○、乙○○支付出國辦理結婚之機票、住宿等費用並提供零用金，再安排戊○○、丙○○、乙○○分別於如附表所示之結婚時間，至如附表所示之地點與傅麗娜、旺施女、彭梅妮辦理註冊結婚，取得結婚證明書，再於附表所示之時間持上開結婚證明書前往附表所示之我國駐外單位公證，待戊○○、丙○○、乙○○返台後，即於如附表所示之時間，在己○○陪同下，攜帶上開經公證之結婚證明書等相關資料，前往如附表所示戶政事務所辦理結婚登記，使不知情之該所承辦之公務員將如附表所示之外籍女子及人頭老公不實之結婚事項，登載於其職務上所掌具有準公文書性質之電腦戶政資料管理檔案內，足以生損害於戶政機關對戶籍管理之正確性（傅麗娜、旺施女、彭梅妮及戊○○、丙○○、乙○○所涉偽造文書犯行，業經臺灣臺中地方法院檢察署檢察官以103年度偵字第11862號為職權不起訴處分確定）。",
            "被告明知陳丙丁死亡後，其遺產屬全體繼承人公同共有，竟罔顧繼承人之權益，利用不知情之成年男子，冒用陳丙丁名義盜領陳丙丁之存款，損及陳丙丁繼承人與合庫銀行對帳戶管理之正確性，所為實有不該；惟念其前無犯罪科刑紀錄，有臺灣高等法院被告前案紀錄表在卷可稽，素行良好，並考量其坦承犯行，且已對告訴人王陳犁雪為部分賠償（詳後述）之犯後態度，兼衡其國中畢業之教育程度，原為越南國籍因與我國人結婚而取得我國國籍之家庭狀況，及其犯罪之動機、目的、手段等一切情狀，量處如主文所示之刑，諭知易科罰金之折算標準",
            "嗣蔡寶緗於103年1月間死亡，辦理繼承登記前，陳泓維於103年7月間向邱濰櫻、陳濰甄表示願給付其二人各100萬元，而將系爭不動產由其一人單獨繼承，邱濰櫻、陳濰甄要求應先將系爭不動產設定抵押權各100萬元予其二人，以擔保日後分割遺產所取得對陳泓維各100萬元之債權。嗣陳泓維於103年10月間欲以系爭不動產設定抵押向銀行借款，因系爭不動產已設定第一順位抵押權予邱濰櫻、陳濰甄，遭銀行評估擔保不足，陳泓維乃向邱濰櫻、陳濰甄表示願清償各100萬元，並徵得其二人同意塗銷系爭抵押權登記，而於103年10月17日偕同代書林淑慧前往臺中市雅潭地政事務所（下稱雅潭地政所）辦理系爭抵押權塗銷登記，陳泓維當場交付面額各80萬元之支票各1張予邱濰櫻、陳濰甄二人，至於未清償之20萬元，則由陳泓維另簽發面額各20萬元之本票各1張作為清償擔保，並承諾待銀行撥款後，會將前開各20萬元部分再設定抵押權予邱濰櫻、陳濰甄，詎陳泓維於邱濰櫻、陳濰甄塗銷系爭抵押權登記後，就前開各20萬元之債務置若罔聞，亦未就前開各20萬元重新設定抵押權予邱濰櫻、陳濰甄，陳濰甄俟將其對陳泓維之20萬元本票債權讓與邱濰櫻，邱濰櫻乃持系爭本票2張，向法院聲請本票強制執行。陳泓維明知其與邱濰櫻係基於上開協議，而邱濰櫻取得系爭20萬元本票2張，並無詐欺之情事，竟意圖使邱濰櫻、陳濰甄受刑事追訴，於104年5月7日16時08分許，至臺灣臺中地方法院檢察署（下稱臺中地檢署）申告欲對邱濰櫻、陳濰甄提出告訴，並扭曲上開事實，申告略以：「被告邱濰櫻、陳濰甄於其母親蔡寶緗死亡後，隱瞞蔡寶緗之自書遺囑，至103年6月17日始寄給告訴人陳泓維該遺囑，並向告訴人陳泓維詐稱系爭不動產先前已向銀行設定抵押借款各100萬元，要求告訴人陳泓維需給付被告二人各100萬元，始願塗銷抵押權辦理遺產分割，將系爭不動產過戶予告訴人陳泓維，惟告訴人陳泓維認為母親蔡寶緗之遺囑只願分配被告各60萬元，而拒絕分配被告二人各100萬元之請求，最後雙方同意由告訴人陳泓維先支付被告二人各80萬元及簽發系爭面額20萬元之本票2張交被告二人收執，待辦理過戶完成後，被告二人應各自再退還告訴人陳泓維40萬元，被告二人另慫恿告訴人陳泓維先與其妻即案外人張欣倫（已改名：余欣倫）辦理離婚，待所有遺產均分配完畢之後，再辦理復婚，以免張欣倫分到其母親蔡寶緗之遺產云云，告訴人陳泓維不疑有他，均同意之，乃交付被告二人各80萬元，另簽發系爭面額各20萬元之本票交被告二人，復於103年6月3日與其妻共同前往臺中市潭子區戶政事務所辦理兩願離婚（嗣相關遺產分割登記完畢之後，陳泓維、張欣倫復於104年1月26日前往前開戶政事務所再次辦理結婚登記復婚）。"]
sample_4 = ["陳泰龍明知其與泰國籍女子陳玉婉（泰國名字為JARUWANWONGPRATOOM，陳玉婉涉嫌本案偽造文書部分，另經臺灣臺南地方法院檢察署檢察官通緝中）間並無結婚之真意，然因詹巧薇、張晉源等人欲找臺灣男子前往泰國與泰國女子辦理假結婚，林清水負責介紹臺灣男子予詹巧薇、張晉源等人以賺取新臺幣（下同）1萬元之仲介費，適陳泰龍依林清水於報紙上刊登之徵人廣告應徵時，由林清水告知泰國女子欲以假結婚之方式申請來臺居留工作，其若擔任假結婚之人頭老公，除不用負擔前往泰國之食宿、機票費用外，尚可獲取5、6萬元之報酬，迨陳泰龍應允同意後，林清水將其轉介紹予詹巧薇、張晉源等人，陳泰龍遂與林清水、詹巧薇、張晉源等人共同基於行使使公務員登載不實公文書之概括犯意聯絡，先由陳泰龍依指示於民國93年4月3日前往泰國，並於同年月7日與陳玉婉在泰國曼谷市法拉坎農區註冊辦事處登記結婚，取得上開註冊辦事處核發之結婚登記書、結婚證書等文件後，陳泰龍即行返臺，並於同年5月19日持經我國駐泰國臺北經濟文化辦事處認證之上開結婚登記書、結婚證書等資料，至臺南縣永康市（改制前稱謂，以下沿用舊制稱謂）戶政事務所填寫結婚登記申請書，申請辦理其與陳玉婉之結婚登記，因此使上開戶政事務所僅有形式審查權之承辦公務員陷於錯誤，於形式上審核後，將陳泰龍與陳玉婉於93年4月7日結婚之不實事項，登載於其職務上所掌屬公文書之戶籍登記資料內，並據以核發戶籍謄本，而足以生損害於戶政機關對於戶政管理及戶籍登記資料維護之正確性",
            "鯂松明知其母張林金英於98年1月5日死亡，張林金英上揭帳戶之存款，係屬遺產之範疇，屬於張林金英之繼承人林張玉雲（長女）、張鯂松（長子）、張朝庚（次子）、張枝傳（三子）、張玉珠（次女）、張長益（四子）、張玉枝（三女）、張玉治（四女）等全體繼承人（以上係按出生先後順序排列）所公同共有，須由全體繼承人填具申請書或同意委任代理人，並檢具相關證件，依據繼承之程序，始得提領款項。詎其竟基於偽造私文書、行使偽造私文書及意圖為自己不法所有之詐欺取財犯意，分別於98年1月6日、1月10日，未經全體繼承人之同意，持張林金英生前交付其保管而持有之前揭存摺及印章等物，至臺中縣太平市（縣市合併後為臺中市太平區○○○路483號太平市農會，在取款憑條上分別填寫「壹拾萬元正」、「陸萬陸仟元正」之提領金額，並於各該取款憑條之存戶簽章欄內盜蓋「張林金英」之印文各1枚，而偽造用以表示係張林金英本人欲向太平市農會提領存款之私文書各1紙，並連同存摺交予各該不知情之太平市農會承辦人員辦理提款事宜而行使之，致太平市農會之承辦人員陷於錯誤，誤以為係張林金英授權張鯂松代為領取存款，遂如數交付上開提領款項予張鯂松，致張鯂松分別詐得張林金英所有上揭帳戶內之存款新臺幣（下同）10萬元及6萬6千元，均足以生損害於張林金英之其餘繼承人林張玉雲、張朝庚、張枝傳、張玉珠、張玉枝、張長益及張玉治等人之權益暨太平市農會對於存戶存款管理之正確性。嗣經林張玉雲、張玉枝查覺上揭帳戶內存款遭盜領後，訴請偵辦，始循線查知上情。",
            "犯罪事實一、陳泓維為邱濰櫻同母異父之兄、陳濰甄同父同母之胞兄，其三人之母親蔡寶緗（已歿）於民國102年10月6日書立遺囑分配遺產表示其所有門牌號碼臺中市○○區○○路0段0000號2樓建物及坐落之臺中市○○區○○段0000地號土地應有部分、門牌號碼臺中市○○區○○路00巷0號等不動產（下稱系爭不動產），如未變賣，由陳泓維及邱濰櫻、陳濰甄三人共同繼承，若嗣後變賣，所得價金由邱濰櫻、陳濰甄二人各分100萬元，其餘均歸陳泓維所有。嗣蔡寶緗於103年1月間死亡，辦理繼承登記前，陳泓維於103年7月間向邱濰櫻、陳濰甄表示願給付其二人各100萬元，而將系爭不動產由其一人單獨繼承，邱濰櫻、陳濰甄要求應先將系爭不動產設定抵押權各100萬元予其二人，以擔保日後分割遺產所取得對陳泓維各100萬元之債權。嗣陳泓維於103年10月間欲以系爭不動產設定抵押向銀行借款，因系爭不動產已設定第一順位抵押權予邱濰櫻、陳濰甄，遭銀行評估擔保不足，陳泓維乃向邱濰櫻、陳濰甄表示願清償各100萬元，並徵得其二人同意塗銷系爭抵押權登記，而於103年10月17日偕同代書林淑慧前往臺中市雅潭地政事務所（下稱雅潭地政所）辦理系爭抵押權塗銷登記，陳泓維當場交付面額各80萬元之支票各1張予邱濰櫻、陳濰甄二人，至於未清償之20萬元，則由陳泓維另簽發面額各20萬元之本票各1張作為清償擔保，並承諾待銀行撥款後，會將前開各20萬元部分再設定抵押權予邱濰櫻、陳濰甄，詎陳泓維於邱濰櫻、陳濰甄塗銷系爭抵押權登記後，就前開各20萬元之債務置若罔聞，亦未就前開各20萬元重新設定抵押權予邱濰櫻、陳濰甄，陳濰甄俟將其對陳泓維之20萬元本票債權讓與邱濰櫻，邱濰櫻乃持系爭本票2張，向法院聲請本票強制執行。陳泓維明知其與邱濰櫻係基於上開協議，而邱濰櫻取得系爭20萬元本票2張，並無詐欺之情事，竟意圖使邱濰櫻、陳濰甄受刑事追訴，於104年5月7日16時08分許，至臺灣臺中地方法院檢察署（下稱臺中地檢署）申告欲對邱濰櫻、陳濰甄提出告訴，並扭曲上開事實，申告略以：「被告邱濰櫻、陳濰甄於其母親蔡寶緗死亡後，隱瞞蔡寶緗之自書遺囑，至103年6月17日始寄給告訴人陳泓維該遺囑，並向告訴人陳泓維詐稱系爭不動產先前已向銀行設定抵押借款各100萬元，要求告訴人陳泓維需給付被告二人各100萬元，始願塗銷抵押權辦理遺產分割，將系爭不動產過戶予告訴人陳泓維，惟告訴人陳泓維認為母親蔡寶緗之遺囑只願分配被告各60萬元，而拒絕分配被告二人各100萬元之請求，最後雙方同意由告訴人陳泓維先支付被告二人各80萬元及簽發系爭面額20萬元之本票2張交被告二人收執，待辦理過戶完成後，被告二人應各自再退還告訴人陳泓維40萬元，被告二人另慫恿告訴人陳泓維先與其妻即案外人張欣倫（已改名：余欣倫）辦理離婚，待所有遺產均分配完畢之後，再辦理復婚，以免張欣倫分到其母親蔡寶緗之遺產云云，告訴人陳泓維不疑有他，均同意之，乃交付被告二人各80萬元，另簽發系爭面額各20萬元之本票交被告二人，復於103年6月3日與其妻共同前往臺中市潭子區戶政事務所辦理兩願離婚（嗣相關遺產分割登記完畢之後，陳泓維、張欣倫復於104年1月26日前往前開戶政事務所再次辦理結婚登記復婚）。"]

斷詞（無去除人名地名）

In [50]:
tokenized_sample_1 = []
for data in sample_1:
    tokenized_sample_1.append(split_passage(data))

In [51]:
tokenized_sample_2 = []
for data in sample_2:
    tokenized_sample_2.append(split_passage(data))

In [66]:
tokenized_sample_3 = []
for data in sample_3:
    tokenized_sample_3.append(split_passage(data))

In [65]:
tokenized_sample_4 = []
for data in sample_4:
    tokenized_sample_4.append(split_passage(data))

In [67]:
print(tokenized_sample_1)
print(tokenized_sample_2)
print(tokenized_sample_3)
print(tokenized_sample_4)

['陳泰 龍 明知 泰國 籍 女子 陳玉 婉 泰國 名字 JARUWANWONGPRATOOM 陳玉 婉 涉嫌 本案 偽造文書 部分 臺灣臺南地方法院檢察署 檢察官 通緝 中 間 無 結婚 真意 張晉 源等 人 欲 找 臺灣 男子 前往 泰國 泰國 女子 辦理 假結婚 林 清水 負責 介紹 臺灣 男子 予 張晉 源等 人 賺取 新臺幣 下同 萬元 仲介 費 適 陳泰 清水 報紙 刊登 人 廣告 應徵 時 清水 告知 泰國 女子 欲 假結婚 方式 申 請來 臺 居留 工作 其若 擔任 假結婚 人頭 不用 負擔 前往 泰國 食宿 機票 費用 外 尚可 獲取 萬元 報酬 迨 陳泰 龍 應允 同意 後 林清 水將 轉 介紹 予 張晉 源等 人 陳泰 龍遂 清水 張晉 源等 人 共同 行使 使 公務員 登載 不實 公文書 概括 犯意 聯絡 先 陳泰 指示 民國 前往 泰國 並於 同年 月 日 陳玉 婉 泰國 法 拉坎 農區 註冊 辦事處 登記 結婚 取得 上開 註冊 辦事處 核發 結婚登記 書 結婚證書 文件 後 陳泰 龍 即行 返臺 並於 同年 持 我國 駐泰國臺北經濟文化辦事處 認證 之上 開 結婚登記 書 結婚證書 資料 臺南縣 永康市 改制 前 稱謂 以下 沿用 舊制 稱謂 戶政事務所 填寫 結婚登記 申請 辦理 陳玉 婉 結婚登記 使 開 戶政事務所 有形 式 審查 承辦 公務員 陷於 錯誤 形式 審核 陳泰 龍 陳玉 婉 結婚 不實 事項 登載 職務 上所 掌屬 公文書 戶籍 登記 資料 核發 戶籍謄本 足以 生 損害 戶政 機關 戶政 管理 戶籍 登記 資料 維護 正確性', '張 慶 堦 係 址 設 宜蘭縣 羅東鎮 路 號 樓 泰雅 工程 有限公司 以下 簡稱 泰雅 公司 實際 負責人 張文 騫 係 址 設 臺中市 區 街 號 工程顧問 有限公司 以下 簡稱 公司 工程師 公司 係 承攬 宜蘭縣政府 文化局 發包 宜蘭國際童玩藝術節 超級 High net 體能 極限 網 設施 委託 監造 服務 案 張文 騫 現場 監工 人員 忠 係 宜蘭縣政府 文化局 約 聘 人員 宜蘭國際童玩藝術節 超級 High net 體能 極限 網 設施 委託 服務 招標 案 承辦 人員 從事 業務 之人 泰雅 公司 民國 間 新臺幣 下同 萬元 得標 承作 宜蘭縣政府 文化局 宜

斷詞（去除人名地名）

In [54]:
cc_tokenized_sample_1 = []
for data in sample_1:
    cc_tokenized_sample_1.append(split_passage(cc.convert(data)))

In [55]:
cc_tokenized_sample_2 = []
for data in sample_2:
    cc_tokenized_sample_2.append(split_passage(cc.convert(data)))

In [69]:
cc_tokenized_sample_3 = []
for data in sample_3:
    cc_tokenized_sample_3.append(split_passage(cc.convert(data)))

In [68]:
cc_tokenized_sample_4 = []
for data in sample_4:
    cc_tokenized_sample_4.append(split_passage(cc.convert(data)))

In [70]:
print(cc_tokenized_sample_1)
print(cc_tokenized_sample_2)
print(cc_tokenized_sample_3)
print(cc_tokenized_sample_4)

['明知 与 籍 女子 名字 为 JARUWANWONGPRATOOM 涉嫌 本案 伪造文书 部分 经 地方法院 检察署 检察官 通缉 中 间 并 无 结婚 真意 源等 人 欲 找 男子 前往 与 女子 办理 假结婚 林 清水 负责 介绍 男子 予 源等 人 赚取 新台币 下同 万元 仲介 费 适 清水 于 报纸 刊登 征 人 广告 应征 时 清水 告知 女子 欲 假结婚 方式 申请 来 台 居留 工作 其若 担任 假结婚 人头 不用 负担 前往 食宿 机票 费用 外 尚可 获取 万元 报酬 迨 应允 同意 后 林 清水 将 转 介绍 予 源等 人 遂 与 清水 源等 人 共同 基于 行使 使 公务员 登载 不实 公文书 概括 犯意 联络 先 指示 于 民国 前往 并 于 同年 月 日 与 法拉 坎农 区 注册 办事处 登记 结婚 取得 上开 注册 办事处 核发 结婚登记 书 结婚证书 文件 后 即行 返台 并 于 同年 持 经 我国 驻 台北 经济 文化 办事处 认证 之上 开 结婚登记 书 结婚证书 资料 永康市 改制 前 称谓 以下 沿用 旧制 称谓 户政事务 填写 结婚登记 申请 办理 与 结婚登记 使 开 户政事务 所仅 有形 式 审查 承办 公务员 陷于 错误 于 形式 审核 后 将 与 于 结婚 不实 事项 登载 于 职务 上所 掌属 公文书 户籍 登记 资料 内 并 据 核发 户籍誊本 足以 生 损害 于 户政 机关 对于 户政 管理 户籍 登记 资料 维护 正确性', '张 庆 堦 系 址 设 镇 路 号 楼 泰雅 工程 有限公司 以下 简称 泰雅 公司 实际 负责人 系址 设 区 街 号 工程顾问 有限公司 以下 简称 公司 工程师 公司 系 承揽 县政府 文化局 发包 国际 艺术节 超级 High net 体能 极限 网 设施 委托 监造 服务 案 为 现场 监工 人员 系 县政府 文化局 约 聘 人员 为 国际 艺术节 超级 High net 体能 极限 网 设施 委托 服务 招标 案 承办 人员 为 从事 业务 之人 泰雅 公司 于 民国 间 新台币 下同 万元 得标 承作 县政府 文化局 国际 艺术节 超级 High net 体能 极限 网 设施 委托 服务 标案 规定 于 投标 时 根据 招标 文件 评选 设施 效果 说明 规范 要求 

## dense

In [29]:
embeddings_1 = model.encode(sentences_1,
                            batch_size=12,
                            max_length=8192, # If you don't need such a long length, you can set a smaller value to speed up the encoding process.
                            )['dense_vecs']
embeddings_2 = model.encode(sentences_2)['dense_vecs']
similarity = embeddings_1 @ embeddings_2.T
print(similarity)

[[0.62590355 0.3474958 ]
 [0.34986818 0.6782462 ]]


In [30]:
embeddings_1 = model.encode(tword_1,
                            batch_size=12,
                            max_length=8192, # If you don't need such a long length, you can set a smaller value to speed up the encoding process.
                            )['dense_vecs']
embeddings_2 = model.encode(tword_2)['dense_vecs']
similarity = embeddings_1 @ embeddings_2.T
print(similarity)

[[0.8963932  0.46727026]
 [0.48538885 0.6649283 ]]


In [31]:
embeddings_1 = model.encode(sword_1,
                            batch_size=12,
                            max_length=8192, # If you don't need such a long length, you can set a smaller value to speed up the encoding process.
                            )['dense_vecs']
embeddings_2 = model.encode(sword_2)['dense_vecs']
similarity = embeddings_1 @ embeddings_2.T
print(similarity)

[[0.8509769  0.5041799 ]
 [0.464958   0.74887514]]


### 斷詞、去除人名地名 測試

#### sample 1 vs. sample 2

有做斷詞 ＆ 去除人名地名 效果最好

In [32]:
embeddings_1 = model.encode(sample_1,
                            batch_size=12,
                            max_length=8192, # If you don't need such a long length, you can set a smaller value to speed up the encoding process.
                            )['dense_vecs']
embeddings_2 = model.encode(sample_2)['dense_vecs']
similarity = embeddings_1 @ embeddings_2.T
print(similarity)

[[0.6247748  0.51102   ]
 [0.4258053  0.55100495]]


In [58]:
embeddings_1 = model.encode(tokenized_sample_1,
                            batch_size=12,
                            max_length=8192, # If you don't need such a long length, you can set a smaller value to speed up the encoding process.
                            )['dense_vecs']
embeddings_2 = model.encode(tokenized_sample_2)['dense_vecs']
similarity = embeddings_1 @ embeddings_2.T
print(similarity)

[[0.6372054  0.5215677 ]
 [0.44952914 0.54799694]]


In [59]:
embeddings_1 = model.encode(cc_tokenized_sample_1,
                            batch_size=12,
                            max_length=8192, # If you don't need such a long length, you can set a smaller value to speed up the encoding process.
                            )['dense_vecs']
embeddings_2 = model.encode(cc_tokenized_sample_2)['dense_vecs']
similarity = embeddings_1 @ embeddings_2.T
print(similarity)

[[0.7014028  0.51494694]
 [0.46641925 0.5667747 ]]


#### sample 3 vs. sample 4
看不出來哪個比較好

In [71]:
embeddings_1 = model.encode(sample_3,
                            batch_size=12,
                            max_length=8192, # If you don't need such a long length, you can set a smaller value to speed up the encoding process.
                            )['dense_vecs']
embeddings_2 = model.encode(sample_4)['dense_vecs']
similarity = embeddings_1 @ embeddings_2.T
print(similarity)

[[0.7075915  0.5663347  0.5656686 ]
 [0.5669893  0.69472134 0.64144325]
 [0.646868   0.62986344 0.9406792 ]]


In [72]:
embeddings_1 = model.encode(tokenized_sample_3,
                            batch_size=12,
                            max_length=8192, # If you don't need such a long length, you can set a smaller value to speed up the encoding process.
                            )['dense_vecs']
embeddings_2 = model.encode(tokenized_sample_4)['dense_vecs']
similarity = embeddings_1 @ embeddings_2.T
print(similarity)

[[0.678797   0.5343967  0.54664844]
 [0.57834435 0.68384033 0.65925336]
 [0.5949729  0.6195218  0.9356131 ]]


In [73]:
embeddings_1 = model.encode(cc_tokenized_sample_3,
                            batch_size=12,
                            max_length=8192, # If you don't need such a long length, you can set a smaller value to speed up the encoding process.
                            )['dense_vecs']
embeddings_2 = model.encode(cc_tokenized_sample_4)['dense_vecs']
similarity = embeddings_1 @ embeddings_2.T
print(similarity)

[[0.73560154 0.5143888  0.57429796]
 [0.54089767 0.6892333  0.66024125]
 [0.57658803 0.64016986 0.9014207 ]]


## sparse

In [35]:
output_1 = model.encode(sentences_1, return_dense=True, return_sparse=True, return_colbert_vecs=False)
output_2 = model.encode(sentences_2, return_dense=True, return_sparse=True, return_colbert_vecs=False)

# you can see the weight for each token:
print(model.convert_id_to_token(output_1['lexical_weights']))
# [{'What': 0.08356, 'is': 0.0814, 'B': 0.1296, 'GE': 0.252, 'M': 0.1702, '3': 0.2695, '?': 0.04092},
#  {'De': 0.05005, 'fin': 0.1368, 'ation': 0.04498, 'of': 0.0633, 'BM': 0.2515, '25': 0.3335}]


# compute the scores via lexical mathcing
lexical_scores = model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_2['lexical_weights'][0])
print(lexical_scores)
# 0.19554901123046875

print(model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_1['lexical_weights'][1]))
# 0.0

[{'What': 0.08362087, 'is': 0.08146954, 'B': 0.12964633, 'GE': 0.25186998, 'M': 0.1700173, '3': 0.26957878, '?': 0.040755123}, {'De': 0.050144203, 'fin': 0.13689385, 'ation': 0.045134626, 'of': 0.06342198, 'BM': 0.25167608, '25': 0.33353215}]
0.19554448779672384
0


In [36]:
output_1 = model.encode(tword_1, return_dense=True, return_sparse=True, return_colbert_vecs=False)
output_2 = model.encode(tword_2, return_dense=True, return_sparse=True, return_colbert_vecs=False)

# you can see the weight for each token:
print(model.convert_id_to_token(output_1['lexical_weights']))
# [{'What': 0.08356, 'is': 0.0814, 'B': 0.1296, 'GE': 0.252, 'M': 0.1702, '3': 0.2695, '?': 0.04092},
#  {'De': 0.05005, 'fin': 0.1368, 'ation': 0.04498, 'of': 0.0633, 'BM': 0.2515, '25': 0.3335}]


# compute the scores via lexical mathcing
lexical_scores = model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_2['lexical_weights'][0])
print(lexical_scores)
# 0.19554901123046875

print(model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_1['lexical_weights'][1]))
# 0.0

[{'': 0.06506694, '老師': 0.3252026}, {'': 0.025284417, '商品': 0.3079706}]
0.004433418158441782
0.0016451796982437372


In [37]:
output_1 = model.encode(sword_1, return_dense=True, return_sparse=True, return_colbert_vecs=False)
output_2 = model.encode(sword_2, return_dense=True, return_sparse=True, return_colbert_vecs=False)

# you can see the weight for each token:
print(model.convert_id_to_token(output_1['lexical_weights']))
# [{'What': 0.08356, 'is': 0.0814, 'B': 0.1296, 'GE': 0.252, 'M': 0.1702, '3': 0.2695, '?': 0.04092},
#  {'De': 0.05005, 'fin': 0.1368, 'ation': 0.04498, 'of': 0.0633, 'BM': 0.2515, '25': 0.3335}]


# compute the scores via lexical mathcing
lexical_scores = model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_2['lexical_weights'][0])
print(lexical_scores)
# 0.19554901123046875

print(model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_1['lexical_weights'][1]))
# 0.0

[{'': 0.062333174, '老师': 0.3212393}, {'': 0.025284417, '商品': 0.3079706}]
0.0033142215106636286
0.0015760579844936728


### 斷詞、去除人名地名 測試

#### sample 1 vs. sample 2
有做斷詞 ＆ 去除人名地名 效果最好

In [81]:
output_1 = model.encode(sample_1, return_dense=True, return_sparse=True, return_colbert_vecs=False)
output_2 = model.encode(sample_2, return_dense=True, return_sparse=True, return_colbert_vecs=False)

print(model.convert_id_to_token(output_1['lexical_weights']))

# compute the scores via lexical mathcing
lexical_scores = model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_2['lexical_weights'][0])
print(lexical_scores)

print(model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_1['lexical_weights'][1]))

[{'陳': 0.2521531, '泰': 0.26584157, '龍': 0.30666173, '明': 0.101543695, '知': 0.12975708, '與': 0.14477229, '泰國': 0.22403938, '籍': 0.15496597, '女子': 0.20520654, '玉': 0.18034828, '婉': 0.29982036, '名字': 0.22417226, '為': 0.11855908, 'JAR': 0.15914416, 'U': 0.13418508, 'WAN': 0.17463517, 'W': 0.10995911, 'ONG': 0.16131966, 'P': 0.064036936, 'RAT': 0.15323746, 'OOM': 0.2604087, '涉嫌': 0.13823399, '本': 0.028688367, '案': 0.15687248, '偽': 0.22356482, '造': 0.1684515, '文書': 0.22656986, '部分': 0.04591082, ',': 0.028518757, '另': 0.012671046, '經': 0.02800768, '臺灣': 0.17963552, '臺': 0.20389023, '南': 0.103920385, '地方': 0.008341521, '法院': 0.055076532, '檢察': 0.06869293, '署': 0.028031692, '官': 0.028371364, '通': 0.014324367, '緝': 0.21278569, '中': 0.027402855, ')': 0.02840618, '間': 0.049174845, '並': 0.06908343, '無': 0.14975798, '結婚': 0.25648907, '之': 0.028479552, '真': 0.07777626, '意': 0.12144834, '然': 0.035248004, '因': 0.014868546, '詹': 0.19410798, '巧': 0.1502397, '薇': 0.24541272, '、': 0.028916981, '張': 0.14631

In [82]:
lexical_scores1 = model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_2['lexical_weights'][0])
lexical_scores2 = model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_2['lexical_weights'][1])

lexical_scores3 = model.compute_lexical_matching_score(output_1['lexical_weights'][1], output_2['lexical_weights'][0])
lexical_scores4 = model.compute_lexical_matching_score(output_1['lexical_weights'][1], output_2['lexical_weights'][1])

print(f'{lexical_scores1} {lexical_scores2}')
print(f'{lexical_scores3} {lexical_scores4}')

0.42963066551601514 0.0953057055012323
0.09472329181153327 0.28883391610725084


In [83]:
output_1 = model.encode(tokenized_sample_1, return_dense=True, return_sparse=True, return_colbert_vecs=False)
output_2 = model.encode(tokenized_sample_2, return_dense=True, return_sparse=True, return_colbert_vecs=False)

print(model.convert_id_to_token(output_1['lexical_weights']))

# compute the scores via lexical mathcing
lexical_scores = model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_2['lexical_weights'][0])
print(lexical_scores)

print(model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_1['lexical_weights'][1]))

[{'陳': 0.23019531, '泰': 0.25232542, '': 0.022680748, '龍': 0.23932248, '明': 0.10540678, '知': 0.15430021, '泰國': 0.22564119, '籍': 0.15457964, '女子': 0.20561604, '玉': 0.18503447, '婉': 0.2801289, '名字': 0.1893617, 'J': 0.09639844, 'ARU': 0.15340573, 'WAN': 0.1527559, 'W': 0.10614167, 'ONG': 0.17406152, 'P': 0.08019862, 'RAT': 0.14316642, 'OOM': 0.25425714, '涉嫌': 0.16655497, '本': 0.022869216, '案': 0.16948897, '偽': 0.23804182, '造': 0.19358818, '文書': 0.24949726, '部分': 0.041564226, '臺灣': 0.1823859, '臺': 0.17620263, '南': 0.1033398, '地方': 0.021239895, '法院': 0.11220011, '檢察': 0.15207851, '署': 0.0019599125, '官': 0.02064361, '通': 0.08354054, '緝': 0.22676866, '中': 0.01608228, '間': 0.032448836, '無': 0.13647482, '結婚': 0.26993293, '真': 0.09816149, '意': 0.022240553, '張': 0.16134399, '晉': 0.21324842, '源': 0.22151151, '等': 0.021718606, '欲': 0.115392976, '找': 0.12489739, '男子': 0.16522005, '前往': 0.12640253, '辦理': 0.14597866, '假': 0.23169234, '林': 0.12794648, '清水': 0.20588462, '負責': 0.010313436, '介紹': 0.1085398

In [84]:
lexical_scores1 = model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_2['lexical_weights'][0])
lexical_scores2 = model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_2['lexical_weights'][1])

lexical_scores3 = model.compute_lexical_matching_score(output_1['lexical_weights'][1], output_2['lexical_weights'][0])
lexical_scores4 = model.compute_lexical_matching_score(output_1['lexical_weights'][1], output_2['lexical_weights'][1])

print(f'{lexical_scores1} {lexical_scores2}')
print(f'{lexical_scores3} {lexical_scores4}')

0.36429209308698773 0.1011430734906753
0.09015482128597796 0.25684958447527606


In [85]:
output_1 = model.encode(cc_tokenized_sample_1, return_dense=True, return_sparse=True, return_colbert_vecs=False)
output_2 = model.encode(cc_tokenized_sample_2, return_dense=True, return_sparse=True, return_colbert_vecs=False)

print(model.convert_id_to_token(output_1['lexical_weights']))

# compute the scores via lexical mathcing
lexical_scores = model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_2['lexical_weights'][0])
print(lexical_scores)

print(model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_1['lexical_weights'][1]))

[{'': 0.05450929, '明': 0.20648107, '知': 0.29140913, '与': 0.28396952, '籍': 0.24085203, '女子': 0.25043726, '名字': 0.19686082, '为': 0.12647548, 'J': 0.13592005, 'ARU': 0.15880142, 'WAN': 0.19843721, 'W': 0.08575899, 'ONG': 0.14353639, 'P': 0.07869031, 'RAT': 0.1598173, 'OOM': 0.27195808, '涉嫌': 0.19210841, '本': 0.017413981, '案': 0.16671188, '伪': 0.2339328, '造': 0.1773647, '文': 0.13739982, '书': 0.20512949, '部分': 0.0564465, '地方': 0.055909, '法院': 0.10825533, '检察': 0.12534922, '官': 0.0072517022, '通': 0.03940476, '缉': 0.24454391, '中': 0.01402327, '间': 0.029456668, '并': 0.03237748, '无': 0.1655582, '结婚': 0.2849247, '真': 0.096296124, '意': 0.13781816, '源': 0.2020495, '等': 0.023157053, '欲': 0.09399369, '找': 0.13135943, '男子': 0.16969621, '前往': 0.11149996, '办理': 0.14429, '假': 0.23981307, '林': 0.15007254, '清水': 0.2241843, '负责': 0.10493385, '介绍': 0.15919706, '予': 0.06341365, '人': 0.03836564, '赚': 0.16588175, '取': 0.01681481, '新': 0.056661643, '台': 0.18780553, '币': 0.15654057, '下': 0.016991924, '同': 0.0254

In [86]:
lexical_scores1 = model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_2['lexical_weights'][0])
lexical_scores2 = model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_2['lexical_weights'][1])

lexical_scores3 = model.compute_lexical_matching_score(output_1['lexical_weights'][1], output_2['lexical_weights'][0])
lexical_scores4 = model.compute_lexical_matching_score(output_1['lexical_weights'][1], output_2['lexical_weights'][1])

print(f'{lexical_scores1} {lexical_scores2}')
print(f'{lexical_scores3} {lexical_scores4}')

0.5111280533892568 0.1395197610836476
0.1275615413323976 0.29369049098386313


#### sample 3 vs. sample 4
看不出來哪個比較好

In [87]:
output_1 = model.encode(sample_3, return_dense=True, return_sparse=True, return_colbert_vecs=False)
output_2 = model.encode(sample_4, return_dense=True, return_sparse=True, return_colbert_vecs=False)

print(model.convert_id_to_token(output_1['lexical_weights']))

# compute the scores via lexical mathcing
lexical_scores = model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_2['lexical_weights'][0])
print(lexical_scores)

print(model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_1['lexical_weights'][1]))

[{'己': 0.24949853, '○○': 0.21181884, '與': 0.092471175, '經營': 0.19112001, '金': 0.17044678, '晉': 0.24103169, '安': 0.20256236, '人力': 0.18579343, '仲介': 0.19294608, '有限公司': 0.15217295, '官': 0.18751574, '威': 0.2310353, '成': 0.21062735, '、': 0.03673291, '怡': 0.21308821, '秀': 0.21219215, '涉': 0.06764519, '違反': 0.19296533, '就業': 0.18222195, '服務': 0.16803238, '法': 0.10056808, '等': 0.024950303, '案件': 0.10885369, ',': 0.024395816, '分別': 0.040344298, '臺灣': 0.19430293, '南投': 0.1477972, '地方': 0.02846387, '法院': 0.11422471, '95': 0.17947875, '年度': 0.089316085, '易': 0.14907512, '字': 0.021837287, '第': 0.048283678, '53': 0.15322298, '號': 0.024714343, '判': 0.1293755, '處': 0.13921583, '有期徒刑': 0.24008499, '10': 0.09736256, '月': 0.024471205, '高等': 0.023227133, '臺': 0.19847867, '中': 0.06706471, '分': 0.1491487, '院': 0.01305303, '以': 0.025113717, '99': 0.15139693, '上': 0.023977656, '94': 0.10927415, '6': 0.07438111, '確定': 0.14778467, '真實': 0.19627538, '姓名': 0.13756905, '年': 0.11792795, '籍': 0.13757485, '不': 0.12

In [88]:
lexical_scores1 = model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_2['lexical_weights'][0])
lexical_scores2 = model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_2['lexical_weights'][1])
lexical_scores3 = model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_2['lexical_weights'][2])

lexical_scores4 = model.compute_lexical_matching_score(output_1['lexical_weights'][1], output_2['lexical_weights'][0])
lexical_scores5 = model.compute_lexical_matching_score(output_1['lexical_weights'][1], output_2['lexical_weights'][1])
lexical_scores6 = model.compute_lexical_matching_score(output_1['lexical_weights'][1], output_2['lexical_weights'][2])

lexical_scores7 = model.compute_lexical_matching_score(output_1['lexical_weights'][2], output_2['lexical_weights'][0])
lexical_scores8 = model.compute_lexical_matching_score(output_1['lexical_weights'][2], output_2['lexical_weights'][1])
lexical_scores9 = model.compute_lexical_matching_score(output_1['lexical_weights'][2], output_2['lexical_weights'][2])

print(f'{lexical_scores1} {lexical_scores2} {lexical_scores3}')
print(f'{lexical_scores4} {lexical_scores5} {lexical_scores6}')
print(f'{lexical_scores7} {lexical_scores8} {lexical_scores9}')

1.0886423303586525 0.3831455946240112 0.38493005066993646
0.27445293439086527 0.402689978201181 0.37952038703951985
0.3369303013168974 0.35157026874560415 2.7369453321980473


In [89]:
output_1 = model.encode(tokenized_sample_3, return_dense=True, return_sparse=True, return_colbert_vecs=False)
output_2 = model.encode(tokenized_sample_4, return_dense=True, return_sparse=True, return_colbert_vecs=False)

print(model.convert_id_to_token(output_1['lexical_weights']))

# compute the scores via lexical mathcing
lexical_scores = model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_2['lexical_weights'][0])
print(lexical_scores)

print(model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_1['lexical_weights'][1]))

[{'經營': 0.23106717, '金': 0.22215083, '人力': 0.19755243, '仲介': 0.2430965, '有限公司': 0.18681578, '官': 0.20664075, '威': 0.25679636, '成': 0.21859004, '涉': 0.12265912, '違反': 0.22415353, '': 0.009732299, '就業': 0.20872307, '服務': 0.22841705, '法': 0.1613704, '案件': 0.13216618, '分別': 0.07402485, '臺灣': 0.1886867, '南投': 0.13048801, '地方': 0.020976283, '法院': 0.13424242, '年度': 0.11446829, '易': 0.16214943, '字': 0.05330687, '號': 0.015390031, '判': 0.136271, '處': 0.11005674, '有期徒刑': 0.24844971, '高等': 0.019260101, '臺': 0.13268411, '中': 0.06684701, '院': 0.02059358, '上': 0.044759147, '確定': 0.14192514, '真實': 0.22802193, '姓名': 0.1665453, '籍': 0.1724526, '不': 0.06680078, '詳': 0.15578593, '綽': 0.20556319, '年': 0.12541798, '女子': 0.18218562, '明': 0.1487738, '知': 0.23481369, '表': 0.011119336, '外籍': 0.19408923, '麗': 0.15711048, '娜': 0.24559414, '工作': 0.16217634, '無': 0.17202279, '結婚': 0.23072706, '真': 0.11060034, '意': 0.12605956, '負責': 0.16488542, '居': 0.11713222, '介紹': 0.052577607, '戊': 0.16554372, '擔任': 0.11732267, '

In [90]:
lexical_scores1 = model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_2['lexical_weights'][0])
lexical_scores2 = model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_2['lexical_weights'][1])
lexical_scores3 = model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_2['lexical_weights'][2])

lexical_scores4 = model.compute_lexical_matching_score(output_1['lexical_weights'][1], output_2['lexical_weights'][0])
lexical_scores5 = model.compute_lexical_matching_score(output_1['lexical_weights'][1], output_2['lexical_weights'][1])
lexical_scores6 = model.compute_lexical_matching_score(output_1['lexical_weights'][1], output_2['lexical_weights'][2])

lexical_scores7 = model.compute_lexical_matching_score(output_1['lexical_weights'][2], output_2['lexical_weights'][0])
lexical_scores8 = model.compute_lexical_matching_score(output_1['lexical_weights'][2], output_2['lexical_weights'][1])
lexical_scores9 = model.compute_lexical_matching_score(output_1['lexical_weights'][2], output_2['lexical_weights'][2])

print(f'{lexical_scores1} {lexical_scores2} {lexical_scores3}')
print(f'{lexical_scores4} {lexical_scores5} {lexical_scores6}')
print(f'{lexical_scores7} {lexical_scores8} {lexical_scores9}')

0.8052082733747739 0.3592659393834765 0.1626352117418719
0.3098407103084355 0.49542216573991027 0.3589866917955078
0.3123540516862704 0.3545336989554926 1.8461087157593283


In [93]:
output_1 = model.encode(cc_tokenized_sample_3, return_dense=True, return_sparse=True, return_colbert_vecs=False)
output_2 = model.encode(cc_tokenized_sample_4, return_dense=True, return_sparse=True, return_colbert_vecs=False)

print(model.convert_id_to_token(output_1['lexical_weights']))

# compute the scores via lexical mathcing
lexical_scores = model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_2['lexical_weights'][0])
print(lexical_scores)

print(model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_1['lexical_weights'][1]))

[{'与': 0.19463667, '经营': 0.25164324, '金': 0.20276935, '人力': 0.18788807, '仲介': 0.221133, '有限公司': 0.17720011, '官': 0.18080811, '威': 0.2193365, '成': 0.17285642, '涉': 0.13828039, '违反': 0.23833296, '就业': 0.22129744, '服务': 0.21554843, '法': 0.17030583, '': 0.03417991, '案件': 0.16019544, '分别': 0.059859477, '经': 0.021375611, '南投': 0.16903314, '地方': 0.022856288, '法院': 0.16138215, '年度': 0.117794536, '易': 0.15559456, '字': 0.03466467, '号': 0.064973645, '判': 0.17420346, '处': 0.15890427, '有期徒刑': 0.27331036, '高等': 0.06475803, '台中': 0.13897598, '分': 0.057719506, '院': 0.027844079, '上': 0.021218225, '确定': 0.19514719, '真实': 0.24822325, '姓名': 0.19814737, '年': 0.15161157, '籍': 0.15664369, '不': 0.109375834, '详': 0.14572164, '绰': 0.15270476, '女子': 0.1964624, '明': 0.1749071, '知': 0.21874839, '附': 0.04485316, '表': 0.021830041, '所示': 0.028266847, '外籍': 0.24199468, '人': 0.065838166, '仅': 0.084658995, '为': 0.021461494, '来': 0.14258012, '台': 0.23128657, '工作': 0.15168023, '无': 0.19748618, '结婚': 0.24578516, '真': 0.138

In [94]:
lexical_scores1 = model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_2['lexical_weights'][0])
lexical_scores2 = model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_2['lexical_weights'][1])
lexical_scores3 = model.compute_lexical_matching_score(output_1['lexical_weights'][0], output_2['lexical_weights'][2])

lexical_scores4 = model.compute_lexical_matching_score(output_1['lexical_weights'][1], output_2['lexical_weights'][0])
lexical_scores5 = model.compute_lexical_matching_score(output_1['lexical_weights'][1], output_2['lexical_weights'][1])
lexical_scores6 = model.compute_lexical_matching_score(output_1['lexical_weights'][1], output_2['lexical_weights'][2])

lexical_scores7 = model.compute_lexical_matching_score(output_1['lexical_weights'][2], output_2['lexical_weights'][0])
lexical_scores8 = model.compute_lexical_matching_score(output_1['lexical_weights'][2], output_2['lexical_weights'][1])
lexical_scores9 = model.compute_lexical_matching_score(output_1['lexical_weights'][2], output_2['lexical_weights'][2])

print(f'{lexical_scores1} {lexical_scores2} {lexical_scores3}')
print(f'{lexical_scores4} {lexical_scores5} {lexical_scores6}')
print(f'{lexical_scores7} {lexical_scores8} {lexical_scores9}')

1.0643429195642966 0.2776009238077677 0.2348859452904435
0.29742844013526337 0.4435920581781829 0.3631424810446333
0.39091309281866415 0.33642640582365857 1.7690151336173585


## colbert

In [41]:
output_1 = model.encode(sentences_1, return_dense=True, return_sparse=True, return_colbert_vecs=True)
output_2 = model.encode(sentences_2, return_dense=True, return_sparse=True, return_colbert_vecs=True)

print(model.colbert_score(output_1['colbert_vecs'][0], output_2['colbert_vecs'][0]))
print(model.colbert_score(output_1['colbert_vecs'][0], output_2['colbert_vecs'][1]))

tensor(0.7797)
tensor(0.4621)


In [42]:
output_1 = model.encode(tword_1, return_dense=True, return_sparse=True, return_colbert_vecs=True)
output_2 = model.encode(tword_2, return_dense=True, return_sparse=True, return_colbert_vecs=True)

print(model.colbert_score(output_1['colbert_vecs'][0], output_2['colbert_vecs'][0]))
print(model.colbert_score(output_1['colbert_vecs'][0], output_2['colbert_vecs'][1]))
print(model.colbert_score(output_1['colbert_vecs'][1], output_2['colbert_vecs'][0]))
print(model.colbert_score(output_1['colbert_vecs'][1], output_2['colbert_vecs'][1]))

tensor(0.9023)
tensor(0.5454)
tensor(0.5274)
tensor(0.6968)


In [43]:
output_1 = model.encode(sword_1, return_dense=True, return_sparse=True, return_colbert_vecs=True)
output_2 = model.encode(sword_2, return_dense=True, return_sparse=True, return_colbert_vecs=True)

print(model.colbert_score(output_1['colbert_vecs'][0], output_2['colbert_vecs'][0]))
print(model.colbert_score(output_1['colbert_vecs'][0], output_2['colbert_vecs'][1]))
print(model.colbert_score(output_1['colbert_vecs'][1], output_2['colbert_vecs'][0]))
print(model.colbert_score(output_1['colbert_vecs'][1], output_2['colbert_vecs'][1]))

tensor(0.8653)
tensor(0.5597)
tensor(0.5299)
tensor(0.7482)


### 斷詞、去除人名地名 測試

#### sample 1 vs. sample 2
有做斷詞 ＆ 去除人名地名 效果最好

In [98]:
output_1 = model.encode(sample_1, return_dense=True, return_sparse=True, return_colbert_vecs=True)
output_2 = model.encode(sample_2, return_dense=True, return_sparse=True, return_colbert_vecs=True)

print(model.colbert_score(output_1['colbert_vecs'][0], output_2['colbert_vecs'][0]),end='')
print(model.colbert_score(output_1['colbert_vecs'][0], output_2['colbert_vecs'][1]))
print(model.colbert_score(output_1['colbert_vecs'][1], output_2['colbert_vecs'][0]),end='')
print(model.colbert_score(output_1['colbert_vecs'][1], output_2['colbert_vecs'][1]))

tensor(0.6411)tensor(0.4766)
tensor(0.5150)tensor(0.6277)


In [96]:
output_1 = model.encode(tokenized_sample_1, return_dense=True, return_sparse=True, return_colbert_vecs=True)
output_2 = model.encode(tokenized_sample_2, return_dense=True, return_sparse=True, return_colbert_vecs=True)

print(model.colbert_score(output_1['colbert_vecs'][0], output_2['colbert_vecs'][0]),end='')
print(model.colbert_score(output_1['colbert_vecs'][0], output_2['colbert_vecs'][1]))
print(model.colbert_score(output_1['colbert_vecs'][1], output_2['colbert_vecs'][0]),end='')
print(model.colbert_score(output_1['colbert_vecs'][1], output_2['colbert_vecs'][1]))

tensor(0.6758)tensor(0.5400)
tensor(0.5783)tensor(0.6691)


In [97]:
output_1 = model.encode(cc_tokenized_sample_1, return_dense=True, return_sparse=True, return_colbert_vecs=True)
output_2 = model.encode(cc_tokenized_sample_2, return_dense=True, return_sparse=True, return_colbert_vecs=True)

print(model.colbert_score(output_1['colbert_vecs'][0], output_2['colbert_vecs'][0]),end='')
print(model.colbert_score(output_1['colbert_vecs'][0], output_2['colbert_vecs'][1]))
print(model.colbert_score(output_1['colbert_vecs'][1], output_2['colbert_vecs'][0]),end='')
print(model.colbert_score(output_1['colbert_vecs'][1], output_2['colbert_vecs'][1]))

tensor(0.7378)tensor(0.5653)
tensor(0.6035)tensor(0.6920)


#### sample 3 vs. sample 4
有做斷詞 ＆ 去除人名地名 效果最好

In [99]:
output_1 = model.encode(sample_3, return_dense=True, return_sparse=True, return_colbert_vecs=True)
output_2 = model.encode(sample_4, return_dense=True, return_sparse=True, return_colbert_vecs=True)

print(model.colbert_score(output_1['colbert_vecs'][0], output_2['colbert_vecs'][0]),end='')
print(model.colbert_score(output_1['colbert_vecs'][0], output_2['colbert_vecs'][1]),end='')
print(model.colbert_score(output_1['colbert_vecs'][0], output_2['colbert_vecs'][2]))
print(model.colbert_score(output_1['colbert_vecs'][1], output_2['colbert_vecs'][0]),end='')
print(model.colbert_score(output_1['colbert_vecs'][1], output_2['colbert_vecs'][1]),end='')
print(model.colbert_score(output_1['colbert_vecs'][1], output_2['colbert_vecs'][2]))
print(model.colbert_score(output_1['colbert_vecs'][2], output_2['colbert_vecs'][0]),end='')
print(model.colbert_score(output_1['colbert_vecs'][2], output_2['colbert_vecs'][1]),end='')
print(model.colbert_score(output_1['colbert_vecs'][2], output_2['colbert_vecs'][2]))

tensor(0.7163)tensor(0.5707)tensor(0.6366)
tensor(0.5765)tensor(0.6438)tensor(0.6344)
tensor(0.7152)tensor(0.7019)tensor(0.9423)


In [100]:
output_1 = model.encode(tokenized_sample_3, return_dense=True, return_sparse=True, return_colbert_vecs=True)
output_2 = model.encode(tokenized_sample_4, return_dense=True, return_sparse=True, return_colbert_vecs=True)

print(model.colbert_score(output_1['colbert_vecs'][0], output_2['colbert_vecs'][0]),end='')
print(model.colbert_score(output_1['colbert_vecs'][0], output_2['colbert_vecs'][1]),end='')
print(model.colbert_score(output_1['colbert_vecs'][0], output_2['colbert_vecs'][2]))
print(model.colbert_score(output_1['colbert_vecs'][1], output_2['colbert_vecs'][0]),end='')
print(model.colbert_score(output_1['colbert_vecs'][1], output_2['colbert_vecs'][1]),end='')
print(model.colbert_score(output_1['colbert_vecs'][1], output_2['colbert_vecs'][2]))
print(model.colbert_score(output_1['colbert_vecs'][2], output_2['colbert_vecs'][0]),end='')
print(model.colbert_score(output_1['colbert_vecs'][2], output_2['colbert_vecs'][1]),end='')
print(model.colbert_score(output_1['colbert_vecs'][2], output_2['colbert_vecs'][2]))

tensor(0.7733)tensor(0.6360)tensor(0.6807)
tensor(0.5986)tensor(0.6798)tensor(0.6493)
tensor(0.7025)tensor(0.7088)tensor(0.9388)


In [101]:
output_1 = model.encode(cc_tokenized_sample_3, return_dense=True, return_sparse=True, return_colbert_vecs=True)
output_2 = model.encode(cc_tokenized_sample_4, return_dense=True, return_sparse=True, return_colbert_vecs=True)

print(model.colbert_score(output_1['colbert_vecs'][0], output_2['colbert_vecs'][0]),end='')
print(model.colbert_score(output_1['colbert_vecs'][0], output_2['colbert_vecs'][1]),end='')
print(model.colbert_score(output_1['colbert_vecs'][0], output_2['colbert_vecs'][2]))
print(model.colbert_score(output_1['colbert_vecs'][1], output_2['colbert_vecs'][0]),end='')
print(model.colbert_score(output_1['colbert_vecs'][1], output_2['colbert_vecs'][1]),end='')
print(model.colbert_score(output_1['colbert_vecs'][1], output_2['colbert_vecs'][2]))
print(model.colbert_score(output_1['colbert_vecs'][2], output_2['colbert_vecs'][0]),end='')
print(model.colbert_score(output_1['colbert_vecs'][2], output_2['colbert_vecs'][1]),end='')
print(model.colbert_score(output_1['colbert_vecs'][2], output_2['colbert_vecs'][2]))

tensor(0.7970)tensor(0.6328)tensor(0.6850)
tensor(0.5972)tensor(0.7019)tensor(0.6748)
tensor(0.6732)tensor(0.7316)tensor(0.9353)


## dense vs. sparse vs. colbert

### sample 1 vs. sample 2
#### dense
  
|          | sample 1 | sample 2 |
| -------- | -------- | -------- |
| sample 1 | 0.7014028| 0.51494694|
| sample 2 | 0.46641925| 0.5667747|

#### sparse

|          | sample 1 | sample 2 |
| -------- | -------- | -------- |
| sample 1 | 0.5111280533892568| 0.1395197610836476|
| sample 2 | 0.1275615413323976| 0.29369049098386313|

#### colbert

|          | sample 1 | sample 2 |
| -------- | -------- | -------- |
| sample 1 | 0.7378| 0.5653|
| sample 2 | 0.6035| 0.6920|


### sample 3 vs. sample 4

#### dense

|          | sample 1 | sample 2 | sample 3 |
| -------- | -------- | -------- | -------- |
| sample 1 | 0.73560154| 0.5143888| 0.57429796|
| sample 2 | 0.54089767| 0.6892333| 0.66024125|
| sample 3 | 0.57658803| 0.64016986| 0.9014207|

#### sparse

|          | sample 1 | sample 2 | sample 3 |
| -------- | -------- | -------- | -------- |
| sample 1 | 1.0643429195642966| 0.2776009238077677| 0.2348859452904435|
| sample 2 | 0.29742844013526337| 0.4435920581781829| 0.3631424810446333|
| sample 3 | 0.39091309281866415| 0.33642640582365857| 1.7690151336173585|

#### colbert

|          | sample 1 | sample 2 | sample 3 |
| -------- | -------- | -------- | -------- |
| sample 1 | 0.7970| 0.6328| 0.6850|
| sample 2 | 0.5972| 0.7019| 0.6748|
| sample 3 | 0.6732| 0.7316| 0.9353|

## 結論
三種方法都可以找出最相關的文章（最相關的文章之間的分數都是最高分），有做斷詞 ＆ 去除人名地名有可能會讓效果變好